### DISCUSSION TITLE
서비스를 CommandService와 QueryService로 분리하는 것에 대해서

### DISCUSSION THREAD

**Discussion by amaran-th** [BDD-CLUB/01-doo-re-back#31] [논의 사항]
안녕하세요! 한 가지 논의하고 싶은 게 생겨 처음으로 디스커션 작성해봅니다ㅎㅎ
현재 저희 두레 프로젝트에서는 도메인마다 Service 클래스를 두고 있습니다.
저는 여기서 Service 클래스를 QueryService, CommandService로 분리하면 좋을 것 같다는 생각이 들었습니다.
데이터를 조회하는 작업은 QueryService에 작성하고, 데이터를 생성/수정/삭제하는 작업은 CommandService에 작성하는 것입니다.(애플리케이션 수준에서 CQRS를 적용한 것)
이렇게 했을 때 장점은 조회 로직과 업데이트 로직을 분리할 수 있어 코드량이 매우 많고 복잡한 도메인의 경우, 단일 클래스로 구현했을 때에 비해 프로덕션 코드와 테스트 코드를 유지보수/관리하기가 좀 더 쉬워진다는 것입니다.
사소하게는 QueryService의 클래스 단에 `@Transactional(readonly=true)`를 붙일 수 있다는 점도 있겠네요

이 안건과 관련하여 논의할 점은 크게 2가지 일 것 같습니다.
1. QueryService/CommandService를 분리하는 전략을 수용할 것인지
2. 수용한다면, 모든 도메인의 서비스에 대해 적용할 것인지, 작성하는 사람이 자유롭게 선택하도록 둘 것인지

의견 많이많이 부탁드립니다~!!

**Comment by shk010130** [comment]
저도 분리하는게 좋아보입니다!

하나의 서비스에 조회 로직과 업데이트 로직을 모두 위치시켰을 때 갈수록 기능과 로직이 많아져서 관리가 힘들었던 기억이 납니다.

**Reply by 02ggang9** [reply]
저도 분리하는게 편하다고 생각합니다~ 과거 책에 이런 내용의 글이 있었는데 보면서 좋다고 생각을 했었습니다! 

과거 프로젝트를 할 때도 클래스 단에 @Transactional(readonly=true) 붙이고 생성/수정/삭제 부분에는 @Transactional을 붙였는데 가끔 실수할 때도 있고 미관상으로도 보기 좋지 않았습니다..

수용한다면 저는 자유보다는 백엔드 컨벤션쪽으로 넣는게 좋다고 생각합니다..!!

**Reply by shkisme** [reply]
컨벤션 동의합니다

**Comment by lcqff** [comment]
좋습니다!

**Comment by JJimini** [comment]
확인

**Comment by amaran-th** [comment]
1. CommandService와 QueryService를 분리하도록 결정
2. 모든 도메인 서비스에 적용하는 것을 컨벤션으로 결정
